src/backend/distributed/commands/multi_copy.c

```diff
@@ -2311,47 +2361,58 @@ CitusSendTupleToPlacements(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest
 	/* connections hash is kept in memory context */
 	MemoryContextSwitchTo(copyDest->memoryContext);
 
-	/* get existing connections to the shard placements, if any */
-	shardConnections = GetShardHashConnections(shardConnectionHash, shardId,
-											   &shardConnectionsFound);
-	if (!shardConnectionsFound)
-	{
-		/*
-		 * Keep track of multi shard accesses before opening connection
-		 * the second shard.
-		 */
-		if (!copyDest->multiShardCopy && hash_get_num_entries(shardConnectionHash) == 2)
-		{
-			Oid relationId = copyDest->distributedRelationId;
-
-			/* mark as multi shard to skip doing the same thing over and over */
-			copyDest->multiShardCopy = true;
+	resetStringInfo(copyOutState->fe_msgbuf);
+	AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+					  copyOutState, columnOutputFunctions, columnCoercionPaths);
 
-			/* error out of conflicting COPY */
-			CheckConflictingParallelCopyAccesses(relationId);
+	shardState = GetShardState(copyDest->shardStateHash,
+							   copyDest->connectionStateHash, shardId);
+	foreach(placementStateCell, shardState->placementStateList)
+	{
+		CopyPlacementState *placementState = lfirst(placementStateCell);
+		CopyConnectionState *connectionState = placementState->connectionState;
+		bool switchOver = false;
 
-			/* when we see multiple shard connections, we mark COPY as parallel modify */
-			RecordParallelModifyAccess(relationId);
+		if (placementState != connectionState->activePlacementState)
+		{
+			if (placementState->data->len > 8 * 1024 * 1024 ||
+				connectionState->activePlacementState == NULL)
+			{
+				switchOver = true;
+			}
+			else
+			{
+				/* buffer data */
+				StringInfo copyBuffer;
+				copyBuffer = copyOutState->fe_msgbuf;
+				appendBinaryStringInfo(placementState->data, copyBuffer->data,
+									   copyBuffer->len);
+			}
 		}
 
-		/* open connections and initiate COPY on shard placements */
-		OpenCopyConnections(copyStatement, shardConnections, stopOnFailure,
-							copyOutState->binary);
+		if (switchOver)
+		{
+			if (connectionState->activePlacementState != NULL)
+			{
+				EndPlacementStateCopyCommand(connectionState->activePlacementState,
+											 copyOutState, binaryCopy);
+				dlist_push_head(&connectionState->bufferedPlacementList,
+								&connectionState->activePlacementState->bufferedPlacementNode);
+			}
+			StartPlacementStateCopyCommand(placementState, copyStatement,
+										   copyOutState, binaryCopy);
+			dlist_delete(&placementState->bufferedPlacementNode);
+			connectionState->activePlacementState = placementState;
+		}
 
-		/* send copy binary headers to shard placements */
-		if (copyOutState->binary)
+		if (placementState == connectionState->activePlacementState)
 		{
-			SendCopyBinaryHeaders(copyOutState, shardId,
-								  shardConnections->connectionList);
+			/* no need to buffer, put data directly on line */
+			SendCopyDataToPlacement(copyOutState->fe_msgbuf, shardId,
```

Comment:
Note that when we just do the switchover the TCP buffer will probably not fit everything. In this case, the data will be buffered in the PGconn and we'll send the next batch the next time `PQflush` is called. We need to make sure the internal buffer doesn't grow too big. In the past, we've had many issues with COPY running out of memory and crashing for this reason.

Reply:
`SendCopyDataToPlacement()` calls `PutRemoteCopyData()` which flushes the internal buffer by calling `FinishConnectionIO()` if internal buffer passes 8MB.

Reply:
True, we should probably add a call to `FinishConnectionIO` to `PutRemoteCopyEnd` as well.

Reply:
`PutRemoteCopyEnd()` already calls `FinishConnecitonIO()`.